cornell-dti/carriage-rider
# Ride flow state and validation refactor

### Summary 






This pull request is the first step towards improving the ride flow state to prevent duplicate data in multiple places (e.g. dates/times in the page state as well as the ride object that was passed between pages, dates/times text controllers in the provider). 
- [x] Removed the Ride object being passed between pages, because this was not getting updated until the proceed button was pressed, which is confusing. RideFlowProvider does have a function for assembling the ride from all of the components, but this is solely used for the summary page, not for the requests, because it can send the individual components of the ride from its state. 
- [x] Pulled all ride creation/editing state into RideFlowProvider, because just keeping state in the pages wouldn't keep the state when the back button is pressed although it would keep the text editing controllers, which caused issues. Just using RideFlowProvider simplifies the request functions and keeps state in one place. 
- [x] Added custom validation logic and rendering for dates/times, because before there was confusing behavior where sometimes times/dates would both update if they were out of order and sometimes only one would update, and the visuals did not match the screen reader text. This was likely due to autovalidate rendering or short-circuit condition evaluation issues that are out of our control.

### Test Plan 

For each of these, confirmed validation errors worked for invalid inputs and switched between recurring and non-recurring flows to check there are issues with state:
- Create regular rides with preset locations, custom locations, invalid custom locations
- Update regular ride
- Update all instances of recurring ride
- Update single instance of recurring ride

Validation expected behavior:
- When first entering a page, no errors
- No errors about empty fields until button is pressed
- If inputted dates/times are out of order, errors appear automatically even without button press
- Fixing inputted dates/times that are out of order removes the errors automatically even without button press
- Screen reader reads the field info, then "error" + the shown error message

### Notes
This new structure affects efficiency because there is more rebuilding due to notifyListeners being called frequently. Ideally, I believe there should be a wrapper RideFlow widget around the ride flow pages which contains all of the state rather than using a Provider to keep a similar organized structure with less rebuilding, but this may also lead to more complex passing of state. I can attempt a refactor to test whether that seems feasible.
